rust_snuba/src/processors/metrics_summaries.rs

```diff
@@ -0,0 +1,158 @@
+use anyhow::Context;
+use rust_arroyo::backends::kafka::types::KafkaPayload;
+use serde::Serialize;
+use uuid::Uuid;
+
+use crate::processors::spans::FromSpanMessage;
+use crate::processors::utils::DEFAULT_RETENTION_DAYS;
+use crate::types::{InsertBatch, KafkaMessageMetadata, RowData};
+
+pub fn process_message(
+    payload: KafkaPayload,
+    _: KafkaMessageMetadata,
+) -> anyhow::Result<InsertBatch> {
+    let payload_bytes = payload.payload().context("Expected payload")?;
+    let msg: FromSpanMessage = serde_json::from_slice(payload_bytes)?;
+
+    if !msg._metrics_summary.is_object() {
+        return Ok(InsertBatch {
+            ..Default::default()
+        });
+    }
+
+    let summaries: MetricsSummaries = msg.try_into()?;
+    let mut rows: Vec<Vec<u8>> = Vec::with_capacity(summaries.metrics_summaries.len());
+
+    for summary in &summaries.metrics_summaries {
+        let serialized = serde_json::to_vec(&summary)?;
+        rows.push(serialized);
+    }
+
+    Ok(InsertBatch {
+        rows: RowData::from_rows(rows),
+        origin_timestamp: None,
+        sentry_received_timestamp: None,
+    })
+}
+
+struct MetricsSummaries {
+    metrics_summaries: Vec<MetricsSummary>,
+}
+
+#[derive(Debug, Default, Serialize)]
+struct MetricsSummary {
+    count: u64,
+    deleted: u8,
+    end_timestamp: u64,
+    max: f64,
+    metric_mri: String,
+    min: f64,
+    project_id: u64,
+    retention_days: u16,
+    span_id: u64,
+    sum: f64,
+    #[serde(rename(serialize = "tags.key"))]
+    tag_keys: Vec<String>,
+    #[serde(rename(serialize = "tags.value"))]
+    tag_values: Vec<String>,
+    trace_id: Uuid,
+}
+
+impl TryFrom<FromSpanMessage> for MetricsSummaries {
+    type Error = anyhow::Error;
+
+    fn try_from(from: FromSpanMessage) -> anyhow::Result<MetricsSummaries> {
+        let mut metrics_summaries: Vec<MetricsSummary> = Vec::new();
+
+        let end_timestamp_ms = from.start_timestamp_ms + from.duration_ms as u64;
+        for (metric_mri, summaries) in from.metrics_summary() {
```

Comment:
nit: if all you need from the span message is the metrics summary, you can define a partial struct that only contains the fields you need to read:

```
struct SpanMessage {
    _metrics_summary: ...
}
```

i think the performance gain should be fairly significant, and then the metrics summaries processor would not require any other fields to be well-formed (ignoring the schema validation)

Reply:
I need a little more than that but that's a better idea.